akka-typed/src/main/scala/akka/typed/Sessions.scala

```diff
@@ -0,0 +1,332 @@
+ *
+ * This means that in order to decouple the Process from the ActorContext
+ * at runtime we must tightly couple the Process DSL to the ActorContext.
+ * It is not yet clear which way to go.
+ *
+ * Sharing Channels between different ActorContexts must be prohibited.
+ * Unfortunately we’d need @local types or similar in order to achive this
+ * at compile-time, so a runtime check will have to suffice.
+ */
+object Sessions {
+
+  /**
+   * This will be a collection of all the management commands that are sent
+   * to a Session-hosting Actor via its main channel (i.e. the ActorRef that
+   * is returned directly for it).
+   *
+   * As an implementation detail this will include Channels, which are sent
+   * as a signal that there is a message pending; this allows zero-allocation
+   * messaging also for the secondary Channels.
+   */
+  trait Command
+
+  /**
+   * Create a Behavior that runs the given Process in a new Actor.
+   */
+  def toProps(p: => Process[Any]): Props[Command] = Props(new ProcessInterpreter(p))
+
+  /**
+   * Trying out the idea of strong coupling between Sessions and ActorContext:
+   * the current context is always available like this, implemented using a
+   * ThreadLocal.
+   */
+  def actorContext: ActorContext[Command] =
+    ProcessImpl.contextHolder.get()(0) match {
+      case null => throw new IllegalStateException("cannot access actorContext outside of the running process")
+      case obj  => obj.asInstanceOf[ActorContext[Command]]
+    }
+
+  /**
+   * Base abstraction modeling a process algebra.
+   */
+  trait Process[+T] { this: ProcessImpl[T] =>
+
+    /**
+     * Sequences that process after this one
+     *
+     * FIXME all implementations below are done in terms of this method => to be optimized
+     */
+    def flatMap[U](f: T => Process[U]): Process[U]
+
+    /**
+     * Sequences that process after this one.
+     */
+    def then[U](other: => Process[U]): Process[U] = flatMap(_ => other)
+
+    /**
+     * Sequences that process after this one.
+     *
+     * This method only exists to accommodate the design choice of Scala
+     * for-comprehensions to rewrite a `yield` clause to `map` and not `flatMap`.
+     */
+    def map[U](f: T => Process[U]): Process[U] = flatMap(f)
+
+    /**
+     * Run the given side-effects after this process completes.
+     */
+    def foreach(f: T => Unit): Process[Unit] = flatMap { t => f(t); unit }
+
+    /**
+     * Evaluate the predicate for this process’ result and abort if it yields `false`.
+     */
+    def filter(pred: T => Boolean): Process[T] = flatMap(t => if (pred(t)) unit(t) else fail)
+
+    /**
+     * Evaluate the predicate for this process’ result and abort if it yields `false`.
+     */
+    def withFilter(pred: T => Boolean): Process[T] = filter(pred)
+
+    /**
+     * Filter and transform this process.
+     */
+    def collect[U](pf: PartialFunction[T, U]): Process[U] = flatMap(t => if (pf.isDefinedAt(t)) unit(pf(t)) else fail)
```

Comment:
Yes, of course. Then I’m left with either the result of the function application or a failure—but I cannot distinguish the failure from one that was returned from the function in order to decide whether it must be wrapped in a unit().